### ISSUE TITLE
Quotas and limits + resource stats

### ISSUE THREAD

**Issue by DiegoPino** [esmero/strawberryfield#150]
# Good afternoon

Hi @giancarlobi and @pcambra 

I'm starting to explore needs and implementations for keeping track of
1. File/storage usage by user/role/collection/object type
2. Allowing Site wide Storage Space/ And ADO count quotas to be set

There are two needs here:
1. To know what is in the site (and most of this can actually be few Solr Queries/Views and + some custom PHP code to sum up/round bytes and done)
2. Permit site admins to have control on how their contributors are using the system. E.g Students, or people providing stories , anonymous submissions (not really, more like authenticated with a limited role)

All the Drupal modules I have seen are quite buggy or 7.x based on not quite well suited. You have some ideas about this? I feel since files are actually entities and we have file sizes we have control over that to some extend (to count) but then enforcing quotas could be handled via the Core Conditions plugin system @pcambra told us about.

See https://www.drupal.org/node/1961370

@alliomeria what use cases needs do you see for this? I'm also thinking of crowd sourced data and Institutional repositories and self deposit of Theses and Homework coming from students

Reporting may want to have some cool https://d3js.org which is a thing I wanted to integrate as a JSON driven formatter too (another ISSUE)

Have you noticed how many issues I have opened lately? I'm on 🔥 🤣

**Comment by giancarlobi** [comment]
@DiegoPino I fully agree that this is a useful feature. Also I agree code has to be based on Solr queries. To do that we need to have all parameters we need to filter on stored into Solr that is into SBF JSON, right? So, i.e. if we need to know how many pages we have then we need for each ADO book type to store number of pages or if we need how many objects per collection we need to store "ismemberof". If this thought is right then do we really need an additional module?

**Comment by pcambra** [comment]
For what we've discussed, I don't think any of the modules available will suit regarding quota, if we're working with "reservations", the issue gets a bit more complex. I kind of like what https://www.drupal.org/project/commerce_stock does with that, having "transactional" stock is a very powerful feature, and they have even an entity type to set where the "stock" comes from (different warehouses have different stock, etc).
Condition plugins + supporting entities depending on the scope would be my take here.
